util/argo/diff.go

```diff
@@ -0,0 +1,379 @@
+package argo
+
+import (
+	"fmt"
+
+	"github.com/go-logr/logr"
+
+	"github.com/argoproj/argo-cd/v2/pkg/apis/application/v1alpha1"
+	appv1 "github.com/argoproj/argo-cd/v2/pkg/apis/application/v1alpha1"
+	"github.com/argoproj/argo-cd/v2/util/argo/managedfields"
+	appstatecache "github.com/argoproj/argo-cd/v2/util/cache/appstate"
+
+	"github.com/argoproj/gitops-engine/pkg/diff"
+	"github.com/argoproj/gitops-engine/pkg/utils/kube"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+)
+
+// DiffConfigBuilder is used as a safe way to create valid DiffConfigs.
+type DiffConfigBuilder struct {
+	diffConfig *diffConfig
+}
+
+// NewDiffConfigBuilder create a new DiffConfigBuilder instance.
+func NewDiffConfigBuilder() *DiffConfigBuilder {
+	return &DiffConfigBuilder{
+		diffConfig: &diffConfig{},
+	}
+}
+
+// WithIgnores sets the list of ResourceIgnoreDifferences in the diff config. Will set an
+// empty list if the input parameter is nil.
+func (b *DiffConfigBuilder) WithIgnores(i []v1alpha1.ResourceIgnoreDifferences) *DiffConfigBuilder {
+	ignores := i
+	if ignores == nil {
+		ignores = []v1alpha1.ResourceIgnoreDifferences{}
+	}
+	b.diffConfig.ignores = ignores
+	return b
+}
+
+// WithOverrides sets the map of ResourceOverride in the diff config. Will set an
+// empty map if the input parameter is nil.
+func (b *DiffConfigBuilder) WithOverrides(o map[string]v1alpha1.ResourceOverride) *DiffConfigBuilder {
+	overrides := o
+	if overrides == nil {
+		overrides = make(map[string]appv1.ResourceOverride)
+	}
+	b.diffConfig.overrides = overrides
+	return b
+}
+
+// WithAppLabelKey sets the appLabelKey in the diff config.
+func (b *DiffConfigBuilder) WithAppLabelKey(l string) *DiffConfigBuilder {
```

Comment:
Nitpick:

Can you consider grouping methods to make it easier to catch errors during compilation time:

* `WithAppLabelKey`, `WithTrackingMethod` -> `WithTracking(appLabelKey string, trackingMethod string)`
* `WithAppName`, `WithNoCache` , `WithStateCache` -> `WithCaching(appName string, appStateCache *appstatecache.Cache, noCache bool)`
* `WithOverrides`, `WithIgnores`, `WithIgnoreAggregatedRoles` -> `WithDiffSettings(o map[string]v1alpha1.ResourceOverride, i []v1alpha1.ResourceIgnoreDifferences, ignoreAggregatedRoles bool)`

I'm hoping this way we will get methods with a manageable number of parameters and code won't compile if we introduce one more "diffing setting"

Reply:
I implemented this approach. I had to make modifications in the validation logic as many of them didn't make sense anymore. Hopefully this will provide us compilation error and better tracking how consumers configure the diff.